File: A2/java-extension/src/Volcano.java
import java.util.ArrayList;

public class Volcano extends Mountain {
	private int countdown;
	private int frequency;
	private boolean active;

	public Volcano(int row, int col, int freq) {
		super(row, col);
		this.countdown = freq;
		this.frequency = freq;
		this.color = "\u001b[41m";
		this.active = true;
	}

	public void act(Map map) {
		this.countdown -= 1;
		if (this.countdown == 0) {
			System.out.println("\033[1;33;41mVolcano erupts! \033[0;0m");
			this.countdown = this.frequency;
			ArrayList<Cell> cells = map.getNeighbours(this.row, this.col);
			for (int i = 0; i < cells.size(); ++i) {
				Object occ = cells.get(i).getOccupant();
				if(occ != null) {
					if (occ instanceof Goblin) {
						((Goblin)occ).setActive(false);
						((Goblin)occ).occupying.removeOccupant();
					} else if (occ instanceof Player) {
						((Player)occ).setHp(((Player)occ).getHp() - 1);
					}
				}
			}
		}
	}

	@Override
	public void display() {
		System.out.format("%s \033[2;97m%d%s \033[0m   ", this.color, this.countdown, this.color);
	}

	// add for java
	public boolean getActive() {
		return this.active;
	}
}

File: A2/python-skeleton/Volcano.py
'''
/*
* CSCI3180 Principles of Programming Languages
*
* --- Declaration ---
*
* I declare that the assignment here submitted is original except for source
* material explicitly acknowledged. I also acknowledge that I am aware of
* University policy and regulations on honesty in academic work, and of the
* disciplinary guidelines and procedures applicable to breaches of such policy
* and regulations, as contained in the website
* http://www.cuhk.edu.hk/policy/academichonesty/
*
* Assignment 2
* Name : Taneja Nikunj
* Student ID : 1155123371
* Email Addr : ntaneja9@cse.cuhk.edu.hk
*/
'''

from Cell import Mountain 

class Volcano(Mountain):
	def __init__(self, row, col, freq):
		Mountain.__init__(self, row, col)
		self._countdown = freq 
		self._frequency = freq
		self._color = '\u001b[41m'
		self._active = True 

	# TODO: _active getter
	@property
	def active(self):
		return self._active
	
	def act(self, map):
		# TODO: reduce the countdown by 1 
        #       when the countdown is zero, refresh the countdown 
        #       get all objects occupying the neighboring cells 
        #		and update their properties accordingly 
		self._countdown -= 1
		if self._countdown == 0: 
			print("\033[1;33;41mVolcano erupts! \033[0;0m")
			# add game logic 
			self._countdown = self._frequency
			cells = map.get_neighbours(self._row, self._col)
			for cell in cells:
				occ = cell.occupant
				if occ:
					if occ.name == "Goblin":
						occ.active = False
						occ.occupying.remove_occupant()
					elif occ.name == "Player":
						occ.hp -= 1
        # END TODO 

	def display(self):
		# TODO: return a string representing the Volcano 
		print(
			f"{self._color} \033[2;97m{self._countdown}{self._color} \033[0m   ",
			end=''
		)
        # END TODO 